### ISSUE TITLE
[DAG] canCreateUndefOrPoison/isGuaranteedNotToBeUndefOrPoison should take an UndefPoisonKind enum

### ISSUE THREAD

**Issue by RKSimon** [llvm/llvm-project#194818]
Currently the DAG variants:
```cpp
bool SelectionDAG::isGuaranteedNotToBeUndefOrPoison(SDValue Op,
                                                    const APInt &DemandedElts,
                                                    bool PoisonOnly,
                                                    unsigned Depth);
bool SelectionDAG::canCreateUndefOrPoison(SDValue Op, const APInt &DemandedElts,
                                          bool PoisonOnly, bool ConsiderFlags,
                                          unsigned Depth);
```
Meaning we can't easily identify an UNDEF element

The ValueTracking & GlobalISel variants use an UndefPoisonKind  enum instead of the PoisonOnly flag:
```
enum class UndefPoisonKind {
  PoisonOnly = (1 << 0),
  UndefOnly = (1 << 1),
  UndefOrPoison = PoisonOnly | UndefOnly,
};
```

They both currently limit scope of this enum, but DAG will likely need it to be exposed in a header to allow the TargetNode callbacks to use it as well.

It might be time to share this enum globally exposed somewhere (Support lib?) to reuse it properly by all 3 implementations

**Comment by levi42x** [comment]
wanna work on this issue.

**Comment by RKSimon** [comment]
> wanna work on this issue.

OK - but be warned, this is not flagged as a good-first-issue on purpose - expect yak shaving and bike shedding!

**Comment by levi42x** [comment]
Understood. I'll take the next two days to dive into the code and work on a solution. If I find I'm unable to make progress by then, I'll let you know so the issue can be unassigned.

**Comment by levi42x** [comment]
Hi @RKSimon. 
Please unassign me from this issue. You were right about the yak shaving—updating the enum triggered a massive chain of dependencies across the different backends that I've hit a wall trying to resolve.

**Comment by RKSimon** [comment]
OK, thanks for trying! To get at least something out of this, what about you create a patch just pulling out the UndefPoisonKind enum into a shared header for the existing use cases?

**Comment by levi42x** [comment]
Understood. 

I'll submit a patch that purely moves the `UndefPoisonKind` enum into `llvm/include/llvm/Support/UndefPoison.h `and updates the existing references in `ValueTracking` and `GlobalISel`.

**Comment by levi42x** [comment]
hi @RKSimon,
I've submitted the minimal patch here: [https://github.com/llvm/llvm-project/pull/195523](https://github.com/llvm/llvm-project/pull/195523). 
This unifies the enum into Support as requested.
